### PR TITLE
Library Tests Branch - Implement Async code using ODM Beanie and async/await 

### PR DESCRIPTION
**Pull request by matanamar10**
add async code

### CODE REVIEW

**Comment by dvirgilad** [comment]
rename this PR.

File: src/controllers/management/borrowing_department.py [hunk 1]
```diff
@@ -13,71 +13,61 @@ class BorrowingDepartment:
...
-        Parameters:
-        - library (Library): The Library object representing the library system.
-        - library_item (LibraryItem): The LibraryItem object (book, disk, etc.) being returned.
-        - patron_id (str): The unique identifier for the patron returning the item.
-
-        Raises:
-        - ValueError: If the patron does not exist, the item is not in the library, the item is not borrowed,
-                      or if the patron has outstanding bills.
+        Handles the asynchronous return process of a library item by a patron.
         """
```
Comment (dvirgilad) [inline src/controllers/management/borrowing_department.py:33]: Why did you delete the doc

File: src/models/entities/library_items/disks/dvd/dvd.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+from typing import Optional
+
+from pydantic import Field
+
+from src.models.entities.library_items.disks.disks import Disk
+
+
+class DVD(Disk):
+    director: str = Field(..., max_length=100, description="Director of the DVD")
```
Comment (dvirgilad) [inline src/models/entities/library_items/disks/dvd/dvd.py]: Why are these basemodel types? The point of an odm is to map the types automatically.

File: src/models/responses/library_items/items.py [hunk 3]
```diff
@@ -26,4 +26,6 @@ class LibraryItemStatusResponse(BaseModel):
 
 class NewItemsResponse(BaseModel):
     message: str = Field(...,
-                         description="A message confirming the successful addition of new items to the library collection")
\ No newline at end of file
+                         description="A message confirming the successful addition of new items to the library collection")
+
+
```
Comment (dvirgilad) [inline src/models/responses/library_items/items.py:35]: remove these spaces

File: src/mongodb/mongodb_models/bills_model.py [hunk 4]
```diff
@@ -1,11 +1,11 @@
...
-    patron_id = StringField(required=True, unique=True, regex=r'^\d{9}$')
-    patron_bill_sum = FloatField(required=True)
-    meta = {
-        'collection': MongoDBSettings().mongo_bills_collection
-    }
+    patron_id: str = Field(..., regex=r'^\d{9}$', unique=True)
+    patron_bill_sum: float = Field(...)
+
+    class Settings:
+        collection = MongoDBSettings().mongo_bills_collection  # Ensure this matches your collection name
```
Comment (dvirgilad) [inline src/mongodb/mongodb_models/bills_model.py]: Remove this comment and stop copying code.
What does this do?

File: src/mongodb/mongodb_models/library_item_model.py [hunk 5]
```diff
@@ -1,10 +1,14 @@
...
 from src.env_utils import MongoDBSettings
 
 
 class LibraryItemDocument(Document):
-    is_borrowed = BooleanField(default=False)
-    title = StringField(required=True, max_length=60)
-    isbn = StringField(required=True, unique=True, regex=r'^\d{9}$')
-    type = StringField(required=True, choices=("Disk", "Book"))
-    meta = {'allow_inheritance': True, 'collection': MongoDBSettings().mongo_items_collection}
+    is_borrowed: bool = Field(default=False)
```
Comment (dvirgilad) [inline src/mongodb/mongodb_models/library_item_model.py]: Do all types have a basemodel type and a document type?

File: src/mongodb/mongo_repository.py [hunk 6]
```diff
@@ -1,143 +1,69 @@
...
-        """
-        Insert a library item document into the database.
-
-        Args:
-            document (LibraryItemDocument): The document to insert.
-        """
-        document.save()
-
-    def delete_document(self, query: Dict[str, Any]) -> None:
-        """
```
Comment (dvirgilad) [inline src/mongodb/mongo_repository.py:37]: Why did you delete these docs?

File: src/mongodb/mongo_repository.py [hunk 7]
```diff
@@ -1,143 +1,69 @@
...
-    def update_item_status(self, isbn: str, is_borrowed: bool) -> None:
-        """
-        Update the status of a library item identified by its ISBN.
-
-        Args:
-            isbn (str): The ISBN of the library item.
-            is_borrowed (bool): True if the item is borrowed, False if it is returned.
-        """
-        item = LibraryItemDocument.objects(isbn=isbn).first()
+class MongoLibraryItemRepository:
```
Comment (dvirgilad) [inline src/mongodb/mongo_repository.py]: Are these classes really needed?
Why not do this functionally

File: src/mongodb/mongo_repository.py [hunk 8]
```diff
@@ -1,143 +1,69 @@
...
-        bill = BillDocument.objects(patron_id=patron_id).first()
-        if bill:
-            bill.update(set__amount=amount)
+            item.is_borrowed = is_borrowed
+            await item.save()
+
+    async def insert_document(self, document: LibraryItemDocument) -> None:
+        await document.insert()
+
+    async def delete_document(self, query: Dict[str, str]) -> None:
```
Comment (dvirgilad) [inline src/mongodb/mongo_repository.py]: these are not indicative names

File: src/mongodb/mongo_setup.py [hunk 9]
```diff
@@ -1,11 +1,26 @@
...
+from src.env_utils import MongoDBSettings
 
+client: AsyncIOMotorClient
 
-# Connect to MongoDB function
-def connect_to_mongodb(settings):
-    connect(db=settings.mongo_db_name, host=settings.mongo_client)
 
+async def connect_to_mongodb():
+    global client
```
Comment (dvirgilad) [inline src/mongodb/mongo_setup.py]: why?

File: src/mongodb/mongo_setup.py [hunk 10]
```diff
@@ -1,11 +1,26 @@
...
+    db = client[MongoDBSettings().mongo_db_name]
+    await init_beanie(database=db, document_models=[LibraryItemDocument, PatronDocument, BillDocument])
+
+
+async def disconnect_from_mongodb():
+    global client
+    if client is None:
+        raise RuntimeError("MongoDB client is not initialized or already disconnected.")
+    client.close()  # Close the client connection (sync method)
+    client = None  # Optionally set client to None after closing
```
Comment (dvirgilad) [inline src/mongodb/mongo_setup.py]: stop copying code
What is this doing?

File: src/routers/library/library.py [hunk 11]
```diff
@@ -1,147 +1,83 @@
...
-
-    This endpoint retrieves the details of a specific library item using its ISBN.
-
-    Args:
-        item_isbn (str): The ISBN of the item to retrieve.
-        library: The library instance, injected via dependency.
-
-    Returns:
-        dict: A dictionary containing the details of the requested library item.
+    Get details of a library item asynchronously.
```
Comment (dvirgilad) [inline src/routers/library/library.py:62]: Stop adding asynchronously to every comment.
I know it is async sync you used `async def`

File: src/main.py [hunk 12]
```diff
@@ -1,19 +1,17 @@
...
 import uvicorn
 
 
 def main():
     try:
-        settings = MongoDBSettings()
-        connect_to_mongodb(settings)
+        connect_to_mongodb()
         app = create_app(library_name="AmarLibrary")
         uvicorn.run(app, host="127.0.0.1", port=8000)
```
Comment (dvirgilad) [inline src/main.py]: You should be getting this from env variables

**Review by dvirgilad — CHANGES_REQUESTED** [review]
You need to read more about what an ODM does.

File: src/mongodb/mongo_setup.py (hunk 9 above)
Comment (matanamar10) [reply src/mongodb/mongo_setup.py]: i dont succeed to connect to mongodb without the global client using async.

File: src/mongodb/mongo_setup.py (hunk 10 above)
Comment (matanamar10) [reply src/mongodb/mongo_setup.py]: i want to close the connection from the mongodb and i define the global client

File: src/main.py (hunk 12 above)
Comment (matanamar10) [reply src/main.py]: i got the MongoDBSettings from env

File: src/controllers/management/borrowing_department.py (hunk 1 above)
Comment (matanamar10) [reply src/controllers/management/borrowing_department.py:33]: While i change the code is deleted in mistake - repeat it from main.

File: src/mongodb/mongo_repository.py (hunk 7 above)
Comment (matanamar10) [reply src/mongodb/mongo_repository.py]: i use it because i got the ABC DAL

File: src/dal/dal.py [hunk 13]
```diff
@@ -10,9 +8,9 @@
 
 class LibraryItemRepository(ABC):
     @abstractmethod
-    def update_item_status(self, isbn: str, is_borrowed: bool) -> None:
+    async def update_item_status(self, isbn: str, is_borrowed: bool) -> None:
         """
-        Update the status of a library item by its ISBN.
+        Asynchronously update the status of a library item by its ISBN.
```
Comment (dvirgilad) [inline src/dal/dal.py]: Stop adding asynchronously everywhere.
Obviously it is async since you used `async def`

File: src/dal/dal.py [hunk 14]
```diff
@@ -84,35 +103,44 @@ def return_item(self, patron_id: str, isbn: str):
         pass
 
     @abstractmethod
-    def insert_document(self, document: PatronDocument) -> None:
+    async def add_patron(self, document: PatronDocument) -> None:
```
Comment (dvirgilad) [inline src/dal/dal.py]: First off, why does this accept a document? What if tomorrow you switch to SQL?
second off, change your parameter name, this function should be generic as well as backend agnostic

File: src/dal/dal.py [hunk 15]
```diff
@@ -84,35 +103,44 @@ def return_item(self, patron_id: str, isbn: str):
...
 
         Args:
-            query (dict): The query to find the document to delete.
+            query (dict): The query to find the document to remove.
         """
         pass
 
     @abstractmethod
-    def patron_exists(self, patron_id: str) -> bool:
+    async def patron_exists(self, patron_id: str) -> bool:
```
Comment (dvirgilad) [inline src/dal/dal.py]: Why is this not called get?
Also why is it not generic?

File: src/dal/dal.py [hunk 16]
```diff
@@ -84,35 +103,44 @@ def return_item(self, patron_id: str, isbn: str):
...
+        Args:
+            patron_id (str): The ID of the patron.
+
+        Returns:
+            bool: True if the patron exists, False otherwise.
+        """
         pass
 
 
 class BillRepository(ABC):
```
Comment (dvirgilad) [inline src/dal/dal.py]: I feel like we talked about this.
The `ABC` needs to be generic (type agnostic) 
read this: https://medium.com/@steveYeah/using-generics-in-python-99010e5056eb

File: src/models/entities/library_items/disks/disks.py [hunk 17]
```diff
@@ -11,4 +11,4 @@
 
 
 class Disk(LibraryItem):
-    type: str = Field(..., max_length=60)
+    disk_type: str = Field(..., max_length=60)
```
Comment (dvirgilad) [inline src/models/entities/library_items/disks/disks.py]: Why did you change this.
The class name does not need to be in a param name.
Same with patron and patron_id

File: src/mongodb/mongo_setup.py (hunk 9 above)
Comment (dvirgilad) [reply src/mongodb/mongo_setup.py]: This is from the beanie docs
```python
async def init():
    # Create Motor client
    client = AsyncIOMotorClient("mongodb://user:pass@host:27017")

    # Init beanie with the Product document class
    await init_beanie(database=client.db_name, document_models=[Product])
```
There should rarely be a need for global variables
https://www.forrestthewoods.com/blog/global-variables-are-evil-and-unsafe/

**Review by dvirgilad — COMMENTED** [review]
This still needs a lot of work

File: src/mongodb/mongodb_models/patron_model.py [hunk 18]
```diff
@@ -1,13 +1,13 @@
-# Define Patron document
-from mongoengine import Document, StringField, DictField, DateTimeField
+from beanie import Document
+from typing import Dict
+from datetime import datetime
 from src.env_utils import MongoDBSettings
 
 
 class PatronDocument(Document):
```
Comment (dvirgilad) [inline src/mongodb/mongodb_models/patron_model.py]: If this is of type Document, why does it need to be named `PatronDocument`?

File: src/controllers/management/finance.py [hunk 19]
```diff
@@ -1,27 +1,29 @@
...
 
     Returns:
         float: The calculated bill amount.
     """
-
     calculated_bill = 0
     check_date = datetime.now()
-    for patron_item_isbn, patron_item_due_date in patron.patron_items.items():
+
+    # Assuming `patron_items` in PatronDocument stores items as a dictionary where the key is the ISBN and value is the due date
```
Comment (dvirgilad) [inline src/controllers/management/finance.py]: Why do you need this comment?
You created the types and decide how state is stored. You don't need to assume anything

File: src/dal/dal.py [hunk 20]
```diff
@@ -1,142 +1,69 @@
...
+from typing import TypeVar, Generic, Type, List, Optional, Dict
 
-from abc import ABC, abstractmethod
-from typing import Optional, Dict, List
-from datetime import datetime
+T = TypeVar("T")
 
-from src.mongodb.mongodb_models.library_item_model import LibraryItemDocument
-from src.mongodb.mongodb_models.patron_model import PatronDocument
+class BaseRepository(Generic[T]):
```
Comment (dvirgilad) [inline src/dal/dal.py:5]: Nice

File: src/dal/dal.py [hunk 21]
```diff
@@ -1,142 +1,69 @@
...
+    """
+    A completely generic base repository for handling common CRUD operations.
+    This class is backend-agnostic and can be extended for any specific database implementation (MongoDB, SQL, etc.).
```
Comment (dvirgilad) [inline src/dal/dal.py:8]: (Assuming the `entity` type implements this interfaces)

**Review by dvirgilad — CHANGES_REQUESTED** [review]
Looking much better.
Left a few small comments
